### DISCUSSION TITLE
大量in-addr.arpa报错 context deadline exceeded

### DISCUSSION THREAD

**Discussion by liangxyz** [IrineSistiana/mosdns#299] [Q&A]
2022-02-21T17:33:02.181Z	warn	mem_cache	cache/cache.go:163	failed to update lazy cache	{"query": "177.20.97.202.in-addr.arpa. IN PTR 18761 1677 192.168.0.36", "error": "forward_local: context deadline exceeded"}
2022-02-21T17:33:02.210Z	warn	main_server	dns_handler/server_handler.go:119	entry returned an err	{"query": "85.102.97.202.in-addr.arpa. IN PTR 35134 1678 192.168.0.36", "error": "hosts: main_sequence: mem_cache: forward_local: context deadline exceeded"}
2022-02-21T17:33:26.090Z	warn	forward_remote	bundled_upstream/bundled_upstream.go:91	upstream failed	{"query": "76.252.55.116.in-addr.arpa. IN PTR 44749 1711 192.168.0.28", "from": "https://dns.233py.com:443/dns-query", "error": "got status code 503 from https://dns.233py.com:443/dns-query"}
2022-02-21T17:33:35.758Z	warn	forward_remote	bundled_upstream/bundled_upstream.go:91	upstream failed	{"query": "66.246.110.36.in-addr.arpa. IN PTR 41942 1733 192.168.0.36", "from": "https://dns.233py.com:443/dns-query", "error": "got status code 503 from https://dns.233py.com:443/dns-query"}

如题，查看mosdns日志，发现大量查询ip.in-addr.arpa.的错误，主要是智能家电进行查询，比如我的360扫地机，此时连局域网的时候是无法通过手机app控制扫地机，点击没反应。断开wifi走流量就能控制了。以下是我的配置：
`log:
  level: info
  file: ''

plugin:

  ################# 服务器插件 ################

  # 启动服务器的插件
  - tag: main_server
    type: server
    args:
      entry:  # 这个服务器插件收到请求后会
        - hosts
        - main_sequence   # 运行主执行序列
        - modify_ttl      # 修改应答的 ttl

      server:   # 监听以下协议和端口
        - protocol: udp
          addr: 0.0.0.0:53
        - protocol: tcp
          addr: 0.0.0.0:53

  ################# 可执行插件 ################
  - tag: hosts
    type: 'hosts'
    args:
      hosts:   # resource record
        - 'ext:/etc/mosdns/hosts'  # 以 `ext:` 开头，从外部文件导入。
#  - tag: arbitrary
#    type: 'arbitrary'
#    args:
#      rr:   # resource record
#        - 'ext:/etc/mosdns/rr.txt'  # 以 `ext:` 开头，从外部文件导入。
  # 包含分流的逻辑的插件
  - tag: main_sequence
    type: sequence
    args:
      exec:
        - if:
            - qtype65
          exec:
            - block
            - _return               # 返回。不再执行后续插件
            
        - if:
            - query_is_ad_domain    # 已知的广告域名
          exec:
            - _block_with_nxdomain  # 生成 NXDOMAIN 应答
            - _return               # 返回。不再执行后续插件。

        - mem_cache      # 运行缓存插件。放在这里就可以避免缓存到无用的广告域名。

        - if:
            - query_is_local_domain   # 已知的本地域名
            - '!_query_is_common'     # 和不常见的请求类型
          exec:
            - forward_local           # 用本地服务器
        
        - if:
            - response_has_ip   # 如果应答包含 IP
          exec:
            - cdn                    # 就直接采用CDN测速结果
            - _return

        - if:
            - query_is_non_local_domain  # 已知的非本地域名
          exec:
            - _prefer_ipv4               # 优先 IPv4
            - forward_remote             # 用远程服务器
        
        - if:
            - response_has_ip   # 如果应答包含 IP
          exec:
            - cdn                    # 就直接采用CDN测速结果
            - _return

        # 剩下的未知域名用 IP 分流。原理类似 ChinaDNS，但更稳定。
        # 详细分流原理请参考 `fallback` 的工作流程。
        - primary:
            - forward_local     # 本地服务器为主。
            
            - if:
                - response_has_ip   # 如果应答包含 IP
              exec:
                - cdn                    # 就直接采用CDN测速结果
            
            - if:
                - '!response_has_local_ip'  # 过滤掉非本地的 IP。
              exec:
                - _drop_response
          secondary:
            - _prefer_ipv4
            - forward_remote
            - if:
                - response_has_ip   # 如果应答包含 IP
              exec:
                - cdn                    # 就直接采用CDN测速结果
          fast_fallback: 200  # 这里建议设置成 local 服务器正常延时的 2~5 倍。
                              # 这个延时保证了 local 延时偶尔变高时，其结果不会被 remote 抢答。
                              # 如果 local 超过这个延时还没响应，可以假设 local 出现了问题。
                              # 这时用就采用 remote 的应答。单位: 毫秒。
          always_standby: true

  # 修改CDN缓存
  - tag: 'cdn'
    type: 'blackhole'
    args:
      ipv4: '162.159.131.143' # 如果请求类型是 A，则生成 IP 为该地址的应答。请修改为自己的CDN测速结果。

  # 缓存
  - tag: 'mem_cache'
    type: 'cache'
    args:
      redis: 'redis://redis:6379/0'
      size: 512000
      lazy_cache_ttl: 259200
      lazy_cache_reply_ttl: 30

  # 修改应答 ttl
  - tag: 'modify_ttl'
    type: 'ttl'
    args:
      minimal_ttl: 300
      maximum_ttl: 3600

  - tag: qtype65
    type: query_matcher
    args:
      qtype: [65]

  - tag: block
    type: blackhole
    args:
      rcode: '0'
      ipv4: '127.0.0.1'
      ipv6: '::1'

  # 转发请求至本地服务器的插件
  - tag: forward_local
    type: forward
    args:
      upstream:
        - addr: 222.172.200.68
        - addr: quic://c.passcloud.xyz:784
        - addr: https://i.233py.com/dns-query

  # 转发请求至远程服务器的插件
  - tag: forward_remote
    type: forward
    args:
      upstream:
        - addr: https://1.1.1.1/dns-query
        - addr: https://dns.233py.com/dns-query
        - addr: quic://c.passcloud.xyz:784

  ################ 匹配器插件 #################

  # 匹配 IP 的插件
  - tag: response_has_ip
    type: response_matcher
    args:
      ip:
        - '1.1.1.0/24'
        - '1.0.0.0/24'
        - '1.1.1.1/32'
        - '1.0.0.1/32'
        - '103.21.244.0/22'
        - '103.22.200.0/22'
        - '103.31.4.0/22'
        - '104.16.0.0/13'
        - '104.24.0.0/14'
        - '108.162.192.0/18'
        - '131.0.72.0/22'
        - '141.101.64.0/18'
        - '162.158.0.0/15'
        - '172.64.0.0/13'
        - '173.245.48.0/20'
        - '188.114.96.0/20'
        - '190.93.240.0/20'
        - '197.234.240.0/22'
        - '198.41.128.0/17'
        - '2606:4700:10::6814:0/112'
        - '2606:4700:10::ac43:0/112'
        - '2606:4700:3000::/48'
        - '2606:4700:3001::/48'
        - '2606:4700:3002::/48'
        - '2606:4700:3003::/48'
        - '2606:4700:3004::/48'
        - '2606:4700:3005::/48'
        - '2606:4700:3006::/48'
        - '2606:4700:3007::/48'
        - '2606:4700:3008::/48'
        - '2606:4700:3009::/48'
        - '2606:4700:3010::/48'
        - '2606:4700:3011::/48'
        - '2606:4700:3012::/48'
        - '2606:4700:3013::/48'
        - '2606:4700:3014::/48'
        - '2606:4700:3015::/48'
        - '2606:4700:3016::/48'
        - '2606:4700:3017::/48'
        - '2606:4700:3018::/48'
        - '2606:4700:3019::/48'
        - '2606:4700:3020::/48'
        - '2606:4700:3021::/48'
        - '2606:4700:3022::/48'
        - '2606:4700:3023::/48'
        - '2606:4700:3024::/48'
        - '2606:4700:3025::/48'
        - '2606:4700:3026::/48'
        - '2606:4700:3027::/48'
        - '2606:4700:3028::/48'
        - '2606:4700:3029::/48'
        - '2606:4700:3030::/48'
        - '2606:4700:3031::/48'
        - '2606:4700:3032::/48'
        - '2606:4700:3033::/48'
        - '2606:4700:3034::/48'
        - '2606:4700:3035::/48'
        - '2606:4700:3036::/48'
        - '2606:4700:3037::/48'
        - '2606:4700:3038::/48'
        - '2606:4700:3039::/48'

  # 匹配本地域名的插件
  - tag: query_is_local_domain
    type: query_matcher
    args:
      domain:
        - 'ext:./geosite.dat:cn'

  # 匹配非本地域名的插件
  - tag: query_is_non_local_domain
    type: query_matcher
    args:
      domain:
        - 'ext:./geosite.dat:geolocation-!cn'

  # 匹配广告域名的插件
  - tag: query_is_ad_domain
    type: query_matcher
    args:
      domain:
        - 'ext:./geosite.dat:category-ads-all'

  # 匹配本地 IP 的插件
  - tag: response_has_local_ip
    type: response_matcher
    args:
      ip:
        - 'ext:./geoip.dat:cn'`

**Comment by IrineSistiana** [comment]
不知道。

ptr 报错看置顶的 issus。

智能设备局域网连不上很可能是你路由防火墙的问题，设备无法互相发现。和 dns 应该没什么关系。
